lib/puppet/ssl/host.rb

```diff
@@ -419,6 +429,140 @@ def load_crls(path)
     end
   end
 
+  # Ensures that the CA certificate is available for either generating or
+  # validating the host's cert.
+  # It will first check if the cert is present in memory (used for testing),
+  # then check on disk, and finally try to download it.
+  # @return true if the CA certificate was found, false otherwise
+  def ensure_ca_certificate
+    file_path = certificate_location('ca')
+    if check_for_certificate_in_memory('ca')
+      return true
+    elsif Puppet::FileSystem.exist?(file_path)
+      load_certificate_bundle(Puppet::FileSystem.read(file_path))
+      return true
+    else
+      bundle = download_ca_certificate_bundle
+      if bundle
+        save_certificate_bundle(bundle)
+        return true
+      end
+    end
+    false
+  end
+
+  # Creates an arry of SSL Certificate objects from a PEM-encoding string
+  # of one or more certs.
+  # @param [String] bundle_string PEM-encoded string of certs
+  # @return [[OpenSSL::X509::Certificate]] the certs loaded from the input string
+  def load_certificate_bundle(bundle_string)
+    delimiters = /-----BEGIN CERTIFICATE-----.*?-----END CERTIFICATE-----/m
+    bundle_string.scan(delimiters).map do |cert|
+      OpenSSL::X509::Certificate.new(cert)
+    end
+  end
+
+  # Fetches the CA certificate bundle from the CA server
+  # @return [[OpenSSL::X509::Certificate]] the certs loaded from the response
+  def download_ca_certificate_bundle
+    return nil if Puppet::SSL::Host.ca_location != :remote
+
+    response = Puppet::Rest::Routes.get_certificate(http_client, 'ca')
+    if response.ok?
+      body = response.read_body
+      load_certificate_bundle(body)
+    else
+      Puppet.warning _('Could not download CA certificate, aborting')
```

Comment:
This warning seems like it should go in the part of code that's actually aborting.